docs/design-documents/2020-02-Promtail-Push-API.md

```diff
@@ -0,0 +1,124 @@
+  labels: {}
+```
+
+### Considerations
+
+Users will be able to define multiple `http` scrape configs, but the base URL
+value must be different for each instance. This allows to cleanly separate
+pipelines through different push endpoints.
+
+Users must also be aware about problems with running Promtail with an HTTP
+target behind a load balancer: if payloads are load balanced between multiple
+promtail instances, ordering of logs in Loki will be disrupted leading to
+rejected pushes. Users are recommended to do one of the following:
+
+1. Have a dedicated Promtail instance for receiving pushes. This also applies to
+   using the syslog target.
+2. Have a separated k8s service that always resolves to the same Promtail pod,
+   bypassing the load balancing issue.
+
+## Alternative Implementation Options
+
+The proposed feature of copying the existing Loki Push API limits how this
+feature can be implemented. The following are a list of alternative
+implementations that allow us to stray from the existing API.
+
+All alternative implementations options assume we will at least have HTTP1
+support.
+
+These options may be combined.
+
+### Option 1: JSON / Protobuf Payload
+
+A new JSON and Protobuf payload format can be designed instead of the existing
+Loki push payload. Both formats would have to be exposed to support clients that
+either can't or won't use protobuf marshalling.
+
+The primary benefit of this approach is to allow us to tweak the payload schema
+independently of Loki's existing schema, but otherwise may not be very useful
+and is essentially just code duplication.
+
+### Option 2: gRPC Service
+
+The
+[logproto.Pusher](https://github.com/grafana/loki/blob/f7ee1c753c76ef63338d53cfba782188a165144d/pkg/logproto/logproto.proto#L8-L10)
+service could be exposed through Promtail. This would enable clients stubs to be
+generated for lanuages that have gRPC support, and, for HTTP1 support, a
+[gRPC Gateway](https://github.com/grpc-ecosystem/grpc-gateway) would be embedded
+in Promtail itself.
+
+This implementation option is similar to the original proposed solution, but
+uses the gRPC gateway to handle HTTP1 traffic instead of the HTTP1 shim that
+Loki uses. There are some concerns with this approach:
+
+1. The gRPC Gateway reverse proxy will need to play nice with the existing HTTP
+   mux used in Promtail.
+2. We couldn't control the HTTP and Protobuf formats separately as Loki can.
+3. Log lines will be double-encoded thanks to the reverse proxy.
+4. A small overhead of using a reverse proxy in-process will be introduced.
+5. This breaks our normal pattern of writing our own shim functions; may add
+   some cognative overhead of having to deal with the gRPC gateway as an outlier
+   in the code.
+
+### Option 3: Plaintext Payload
+
+Prometheus' [Push Gateway API](https://github.com/prometheus/pushgateway#command-line)
+is cleverly designed and we should consider implementing our API in the same
```

Comment:
The Push Gateway has been designed only for difficult metrics collection. Not to support big throughput of metrics. Batching is and will always perform better than this. 

This is probably why we don't use this API in cortex to push metrics.

Reply:
Yeah, I agree it's a bad primary approach for the lack of throughput.